multidialer.go

```diff
@@ -5,12 +5,13 @@ import (
 	"fmt"
 	"math/rand"
 	"net"
+
+	"github.com/libp2p/go-netroute"
 )
 
 type multiDialer struct {
-	loopback    []*net.TCPAddr
-	unspecified []*net.TCPAddr
```

Comment:
Why get rid of these? Without them, we end up re-creating these lists on every dial.

Reply:
The main change is that we've not got an iteration through all listening addresses we need to do on dial now to check if any of them match the preferred source coming out of netroute.

Reply:
Yes, but we don't also need to loop through the unspecified and loopback ones, collecting them into an array. My objection is that we're doing _exactly_ what we did before (collect all loopback/unspecified addresses into arrays) but we're doing it on every dial instead of up-front.